src/sentry/integrations/bitbucket_server/repository.py

```diff
@@ -0,0 +1,113 @@
+from __future__ import absolute_import
+
+import six
+
+from django.core.urlresolvers import reverse
+from sentry.models.integration import Integration
+from sentry.plugins.providers.integration_repository import IntegrationRepositoryProvider
+from sentry.utils.http import absolute_uri
+from sentry.integrations.exceptions import ApiError, IntegrationError
+
+
+class BitbucketServerRepositoryProvider(IntegrationRepositoryProvider):
+    name = "Bitbucket Server"
+
+    def get_installation(self, integration_id, organization_id):
+        if integration_id is None:
+            raise IntegrationError("Bitbucket Server requires an integration id.")
+        integration_model = Integration.objects.get(
+            id=integration_id, organizations=organization_id, provider="bitbucket_server"
+        )
+
+        return integration_model.get_installation(organization_id)
+
+    def get_repository_data(self, organization, config):
+        installation = self.get_installation(config.get("installation"), organization.id)
+        client = installation.get_client()
+        try:
+            parts = config["identifier"].split("/")
+            repo = client.get_repo(parts[0], parts[1])
+        except Exception as e:
+            installation.raise_error(e)
+        else:
+            config["external_id"] = six.text_type(repo["id"])
+            config["name"] = repo["project"]["name"] + "/" + repo["name"]
+        return config
+
+    def build_repository_config(self, organization, data):
+        installation = self.get_installation(data.get("installation"), organization.id)
+        client = installation.get_client()
+        parts = data["identifier"].split("/")
+
+        try:
+            resp = client.create_hook(
+                parts[0],
+                parts[1],
+                {
+                    "name": "sentry-bitbucket-server-repo-hook",
+                    "url": absolute_uri(
+                        reverse(
+                            "sentry-extensions-bitbucketserver-webhook",
+                            kwargs={
+                                "organization_id": organization.id,
+                                "integration_id": data.get("installation").encode("utf-8"),
+                            },
+                        )
+                    ),
+                    "active": True,
+                    "events": ["repo:refs_changed", "pr:merged"],
```

Comment:
@pierredup I see for the BitBucket Cloud integration we have hooks for the following events:
```
["repo:push", "pullrequest:fulfilled"]
```
Are they the same events in BitBucket server with different names?

Reply:
Yes it's the same events, but BitBucket server just has different names